File: materials/glass.py
from materials.material import *

class Glass(Material):
    def __init__(self, color=Color(1, 1, 1), idx_refraction=1.5):
        super().__init__(color)
        self.idx_refraction = idx_refraction

    def bounce(self, r, normal, hit_pos):
        outside = r.direction.dot(normal) < 0
        refraction_ratio = self.idx_refraction if not outside else 1 / self.idx_refraction

        normal = normal if outside else -1 * normal

        unit_direction = r.direction.normalize()
        cos_theta = min(-unit_direction.dot(normal), 1)
        sin_theta = sqrt(1 - cos_theta**2)

        cannot_refract = refraction_ratio * sin_theta > 1
        if cannot_refract or Glass.reflectance(cos_theta, refraction_ratio) > random():
            new_direction = Glass.reflect(unit_direction, normal)
        else:
            new_direction = Glass.refract(
                r.direction, normal, refraction_ratio)
        return Ray(hit_pos, new_direction)

    @staticmethod
    def refract(direction, normal, refraction_ratio):
        cos_theta = min((-1 * direction).dot(normal), 1)
        perp = refraction_ratio * (direction + cos_theta * normal)
        parallel = -sqrt(abs(1 - perp.length()**2)) * normal

        return parallel + perp

    @staticmethod
    def reflect(direction, normal):
        return direction - 2 * direction.dot(normal) * normal

    @staticmethod
    def reflectance(cos, reflectance_ratio):
        r0 = (1 - reflectance_ratio) / (1 + reflectance_ratio)
        r0 = r0**2
        return r0 + (1 - r0) * pow((1 - cos), 5)

File: materials/glass.cpp
#include "glass.h"

Glass::Glass(const Vector &color, const double ior)
    : Material(color), ior(ior) {}

Ray Glass::bounce(const Ray &r, const Vector &normal,
                  const Vector &hitPos) const {
    auto outside = r.direction.dot(normal) < 0;
    auto refractionRatio = outside ? 1.0 / ior : ior;

    auto _normal = outside ? normal : -normal;

    auto unitDirection = r.direction.normalize();
    auto cosTheta = fmin(-unitDirection.dot(_normal), 1.0);
    auto sinTheta = sqrt(1.0 - cosTheta * cosTheta);

    auto cannotRefract = refractionRatio * sinTheta > 1.0;
    double randomNum = random();
    auto newDirection =
        cannotRefract || reflectance(cosTheta, refractionRatio) > randomNum
            ? reflect(unitDirection, _normal)
            : refract(unitDirection, _normal, refractionRatio);

    return Ray(hitPos, newDirection);
}

Vector Glass::refract(const Vector &direction, const Vector &normal,
                      const double refractionRatio) {
    auto cosTheta = fmin(-direction.dot(normal), 1.0);
    Vector rOutPerp = refractionRatio * (direction + cosTheta * normal);
    Vector rOutParallel = -sqrt(fabs(1.0 - rOutPerp.lengthSquared())) * normal;
    return rOutPerp + rOutParallel;
}

Vector Glass::reflect(const Vector &direction, const Vector &normal) {
    return direction - 2 * direction.dot(normal) * normal;
}

double Glass::reflectance(double cosine, double refractionRatio) {
    auto r0 = (1 - refractionRatio) / (1 + refractionRatio);
    r0 = r0 * r0;
    return r0 + (1 - r0) * pow((1 - cosine), 5);
}